text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,786 @@
+|               │        Done         │                     |
+|               └─────────────────────┘                     |
+└───────────────────────────────────────────────────────────┘
+```
+
+## Construct Upgrade Strategies
+
+As we've seen from the above [Working Backward](#working-backward) examples and the [Visual Workflow](#working-backward-visual-workflow),
+there are two strategies employed to execute and deploy the upgraded construct stack: CDK Import strategy and CDK Refactor
+strategy.
+
+### CDK Refactor Strategy
+
+For construct upgrades that maintain the same underlying CloudFormation resource types, CDK users can leverage the CDK
+refactor feature, which uses CloudFormation stack refactoring under the hood. This strategy works because:
+
+- The underlying CloudFormation resource types remain the same (e.g., AWS::EC2::VPC)
+- No additional resources need to be created or deleted
+
+The refactor strategy is safer because CloudFormation stack refactoring operations don't allow new resource creations,
+resource deletions, or changes to resource configurations.
+
+### CDK Import strategy
+
+For construct upgrades that involve changing the underlying resource type (e.g., `AWS::DynamoDB::Table` to `AWS::DynamoDB::GlobalTable`)
+or removing custom resources (e.g., DynamoDB replica custom resources), CDK users need to use the CDK import strategy.
+
+This strategy requires:
+
+- Setting `RemovalPolicy.RETAIN` on the original resources to prevent accidental deletion
+- Specifying physical names (e.g., `tableName`) to import existing resources
+
+## CI/CD Integration
+
+Enterprise developers often lack production AWS credentials, making pre-deployment
+validations (e.g., drift detection, change set analysis) impossible in production environments.
+
+At development time, here is how it works for common enterprise scenarios. Assuming developers have upgrade
+their CDK stack code to use new construct. They can run the following commands to validate changes.
+
+```sh
+$ cdk drift --rejectAWS::DynamoDB::Table
+$ cdk diff --retain AWS::DynamoDB::Table,Custom::DynamoDBReplica --import AWS::DynamoDB::GlobalTable
+```
+
+Depending on the deployment workflow and environment constraints, there are two common enterprise
+patterns:
+
+1. **CI/CD-driven Deployment**:
+For teams with access to CI/CD and production environments, developers commit the upgraded CDK stack
+code to version control. The CI/CD pipeline runs the following as part of the deployment process:
+
+```sh
+# Run validations and store exit code
+cdk drift --rejectAWS::DynamoDB::Table && cdk diff --retain AWS::DynamoDB::Table,Custom::DynamoDBReplica --import AWS::DynamoDB::GlobalTable
+
+if [ $? -eq 0 ]; then
+    echo "Validation passed. Proceeding with deployment..."
+    cdk deploy --import-existing-resources # `cdk refactor --unstable=refactor` for VPC constructs upgrade
+else
+    echo "Validation failed. Deployment blocked."
+    exit 1
+fi
+```
```

Comment:
Can you be specific about the entire workflow end to end? I think it looks like this:

1. Clear their deployment pipelines, no change may be flowing through it
2. Update their pipeline definition to put the above command in it
3. Deploy a single change, the migration change
4. Wait for the change to clear the pipeline (no other changes may go through it)
5. Update their pipeline definition to remove the above command

Is that right?

----

Now think of a typical AWS pipeline that has ~3 pre-prod stages and >20 prod stages, and various approval workflows. Does the above still sound acceptable?

Reply:
I think we should say "CI/CD" is too hard to design right now, we will do it later.

Reply:
I think CI/CD is one of the core feature for construct upgrade. We don't have to fully support complex setup but we should definitely explain the workflow for simpler setup.

I agree the workflow you mentioned doesn't seem ideal especially when users want to intermix construct upgrade with normal feature release changes. The issue with AWS pipeline is that there's usually a 12 hours bake time in between each stage which makes things worse. In either these two case,s I think we can add a note to the RFC to explain the difficulty and suggest users to use the Ops-Lead workflow instead.